### PR TITLE
fix(plugin:douyutv): fix issue 2943

### PR DESCRIPTION
**Pull request by hxzhao527**
fix issue #2943 

import new dep [js2py](https://github.com/PiotrDabkowski/Js2Py)

NO py2 support

### CODE REVIEW

File: src/streamlink/plugins/douyutv.py [hunk 1]
```diff
@@ -1,25 +1,26 @@
+import hashlib
 import re
 import time
-import hashlib
 
-from requests.adapters import HTTPAdapter
+import js2py
```
Comment (back-to) [inline src/streamlink/plugins/douyutv.py:5]: adding a new dependency is bad

**Comment by Erk-** [comment]
I think that we could look into is if the js is static in any way, Last time I looked at it they sent what seemed to be a lot different js. we could probably dump it and see if there is a difference between the code from channel to channel or connection to connection.

**Comment by hxzhao527** [comment]
Old: https://web.archive.org/web/20200502075004/https://www.douyu.com/9999  
Now: https://www.douyu.com/9999  
we can see the sign function `ub98484234` have different content.  
and the value returned is not same

**Comment by Erk-** [comment]
If you try a bunch of streams? Will the sign function be similar or seemingly random?

**Comment by hxzhao527** [comment]
The sign function seems updated daily. But it is same between  different rooms.

If importing new dep is not allowed, we have to make a new release when douyu update.  
Or implement a transformer to translate the js code by ourself.

**Comment by Erk-** [comment]
Could you try and capture a few days of sign code so we can diff them and see if there is any pattern we can use. 

I do agree that the easiest way would be to use something like js2py or similar though.

**Comment by steven7851** [comment]
I think you can create a JS engine for plugin:douyutv. Like [ykdl](https://github.com/zhangn1985/ykdl/blob/master/ykdl/util/jsengine.py) have their own one.

**Comment by Erk-** [comment]
That would add a external depedency, which may or may not be preffered.

**Comment by hxzhao527** [comment]
I have captured three days, and upload the result to https://github.com/hxzhao527/4Issue/tree/master/streamlink-plugin-douyu

here is the diff result with vscode
image
image

**Comment by Erk-** [comment]
I cannot really see if there is a way to make a generic implementation, it looks like it is doing some kind of hash and/or encryption. So I think it will be up to the maintainers how to continue from here.

**Pull request closed without merging**